bndtools.core/src/org/bndtools/core/editors/quickfix/BuildpathQuickfixProcessor.java

```diff
@@ -0,0 +1,165 @@
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.compiler.IProblem;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ImportDeclaration;
+import org.eclipse.jdt.core.dom.Name;
+import org.eclipse.jdt.ui.text.java.IInvocationContext;
+import org.eclipse.jdt.ui.text.java.IJavaCompletionProposal;
+import org.eclipse.jdt.ui.text.java.IProblemLocation;
+import org.eclipse.jdt.ui.text.java.IQuickFixProcessor;
+
+import aQute.bnd.build.Container;
+import aQute.bnd.build.Project;
+import aQute.bnd.build.Workspace;
+import aQute.bnd.osgi.BundleId;
+import aQute.lib.exceptions.Exceptions;
+import bndtools.Plugin;
+import bndtools.central.Central;
+
+public class BuildpathQuickfixProcessor implements IQuickFixProcessor {
+
+	@Override
+	public boolean hasCorrections(ICompilationUnit unit, int problemId) {
+		switch (problemId) {
+			case IProblem.IsClassPathCorrect :
+				return true;
+			case IProblem.ImportNotFound :
+				return true;
+			case IProblem.UndefinedType :
+				return true;
+			default :
+				return false;
+		}
+	}
+
+	@Override
+	public IJavaCompletionProposal[] getCorrections(IInvocationContext context, IProblemLocation[] locations)
+		throws CoreException {
+		try {
+			List<IJavaCompletionProposal> proposals = new ArrayList<>();
+
+			ICompilationUnit compUnit = context.getCompilationUnit();
+			IJavaProject java = compUnit.getJavaProject();
+			if (java == null)
+				return null;
+
+			Project project = Central.getProject(java.getProject());
+			if (project == null)
+				return null;
+
+			boolean test = isInDir(project.getTestSrc(), compUnit.getResource());
+			Workspace workspace = project.getWorkspace();
+
+			for (IProblemLocation location : locations) {
+
+				if (!hasCorrections(context.getCompilationUnit(), location.getProblemId()))
+					return null;
+
+				String partialClassName = getPartialClassName(location.getCoveringNode(context.getASTRoot()));
+				if (partialClassName == null && location.getProblemArguments().length > 0)
+					partialClassName = location.getProblemArguments()[0];
+
+				if (partialClassName == null)
+					continue;
+
+				Map<String, List<BundleId>> result = workspace.search(partialClassName)
+					.orElseThrow(s -> new CoreException(new Status(IStatus.ERROR, Plugin.PLUGIN_ID, s)));
+
+				Set<BundleId> buildpath = getBundleIds(project.getBuildpath());
+				Set<BundleId> testpath = test ? getBundleIds(project.getTestpath()) : Collections.emptySet();
+
+				result.entrySet()
+					.forEach(e -> {
+						for (BundleId id : e.getValue()) {
+
+							if (test && !testpath.contains(id) && !buildpath.contains(id))
+								proposals.add(propose(e.getKey(), id, context, location, project, "-testpath"));
+
+							if (!buildpath.contains(id))
+								proposals.add(propose(e.getKey(), id, context, location, project, "-buildpath"));
+
+						}
+					});
+			}
+			return proposals.toArray(new IJavaCompletionProposal[0]);
```

Comment:
For the common case where we haven't found any proposals, perhaps we should return `null` rather than an empty array. Saves the overhead of the extra object allocations. Relatively small optimisation though.

Reply:
less is more ...

Reply:
Having re-read the spec for `getCorrections()` (per the javadoc), just to let you know that this is technically non-compliant - the javadoc calls for `null` to be returned if you don't have any proposals. It will probably work, but we might one day run into an issue if Eclipse ever relies on the assumption that it should be `null`.

Reply:
I disagree but it is not worth to fight about this. Fixed.